lib/topics_filter.rb

```diff
@@ -388,6 +419,138 @@ def filter_by_number_of_views(min: nil, max: nil)
     filter_by_topic_range(column_name: "topics.views", min:, max:)
   end
 
+  # users:a,b => any of a or b participated in the topic
+  # users:a+b => both a and b participated in the topic
+  def filter_users(values:)
+    values.each do |value|
+      if value.include?("+")
+        usernames = value.split("+")
+        require_all = true
+        if value.include?(",")
+          # no mix and match
+          @scope = @scope.none
+          next
+        end
+      else
+        usernames = value.split(",")
+        require_all = false
+        if value.include?("+")
+          # no mix and match
+          @scope = @scope.none
+          next
+        end
+      end
+
+      usernames = usernames.map(&:downcase).reject(&:blank?)
+
+      if usernames.empty?
+        @scope = @scope.none
+        next
+      end
+
+      user_ids = User.where(staged: false).where("username_lower IN (?)", usernames).pluck(:id)
+
+      if user_ids.empty?
+        @scope = @scope.none
+        next
+      end
+
+      if require_all
+        if user_ids.length < usernames.length
+          @scope = @scope.none
+          next
+        end
+
+        # A possible alternative is to select the topics with the users with the least posts
+        # then expand to all of the rest of the users, this can limit the scanning
+        user_ids.each_with_index { |uid, idx| @scope = @scope.where(<<~SQL) }
+            EXISTS (
+              SELECT 1
+              FROM posts p#{idx}
+              WHERE p#{idx}.topic_id = topics.id AND p#{idx}.user_id = #{uid}
+              LIMIT 1
+            )
+          SQL
+      else
+        @scope = @scope.where(<<~SQL, user_ids: user_ids)
+              topics.id IN (
+                SELECT DISTINCT p.topic_id
+                FROM posts p
+                WHERE p.user_id IN (:user_ids)
+              )
+            SQL
+      end
+    end
+  end
+
+  # group:staff,moderators => any of the groups have participation
+  # group:staff+moderators => both groups have participation
+  def filter_groups(values:)
+    values.each do |value|
```

Comment:
I think this `values.each` code could be split out into a `calculate_all_or_any_scope` method or something, since it does the exact same thing for groups and users